compiler/tests/evm_contracts.rs

```diff
@@ -249,43 +249,51 @@ fn test_assert() {
 }
 
 #[rstest(fixture_file, input, expected,
-    case("while_loop_with_continue.fe", vec![], Some(u256_token(1))),
-    case("while_loop.fe", vec![], Some(u256_token(3))),
-    case("while_loop_with_break.fe", vec![], Some(u256_token(1))),
-    case("while_loop_with_break_2.fe", vec![], Some(u256_token(1))),
-    case("if_statement.fe", vec![6], Some(u256_token(1))),
-    case("if_statement.fe", vec![4], Some(u256_token(0))),
-    case("if_statement_2.fe", vec![6], Some(u256_token(1))),
-    case("if_statement_with_block_declaration.fe", vec![], Some(u256_token(1))),
-    case("ternary_expression.fe", vec![6], Some(u256_token(1))),
-    case("ternary_expression.fe", vec![4], Some(u256_token(0))),
-    case("call_statement_without_args.fe", vec![], Some(u256_token(100))),
-    case("call_statement_with_args.fe", vec![], Some(u256_token(100))),
-    case("call_statement_with_args_2.fe", vec![], Some(u256_token(100))),
+    case("while_loop_with_continue.fe", vec![], Some(uint_token(1))),
+    case("while_loop.fe", vec![], Some(uint_token(3))),
+    case("while_loop_with_break.fe", vec![], Some(uint_token(1))),
+    case("while_loop_with_break_2.fe", vec![], Some(uint_token(1))),
+    case("if_statement.fe", vec![6], Some(uint_token(1))),
+    case("if_statement.fe", vec![4], Some(uint_token(0))),
+    case("if_statement_2.fe", vec![6], Some(uint_token(1))),
+    case("if_statement_with_block_declaration.fe", vec![], Some(uint_token(1))),
+    case("ternary_expression.fe", vec![6], Some(uint_token(1))),
+    case("ternary_expression.fe", vec![4], Some(uint_token(0))),
+    case("call_statement_without_args.fe", vec![], Some(uint_token(100))),
+    case("call_statement_with_args.fe", vec![], Some(uint_token(100))),
+    case("call_statement_with_args_2.fe", vec![], Some(uint_token(100))),
     case("return_bool_true.fe", vec![], Some(bool_token(true))),
     case("return_bool_false.fe", vec![], Some(bool_token(false))),
-    case("return_u256_from_called_fn_with_args.fe", vec![], Some(u256_token(200))),
-    case("return_u256_from_called_fn.fe", vec![], Some(u256_token(42))),
-    case("return_u256.fe", vec![], Some(u256_token(42))),
-    case("return_identity_u256.fe", vec![42], Some(u256_token(42))),
+    case("return_u256_from_called_fn_with_args.fe", vec![], Some(uint_token(200))),
+    case("return_u256_from_called_fn.fe", vec![], Some(uint_token(42))),
+    case("return_u256.fe", vec![], Some(uint_token(42))),
+    // case("return_u128.fe", vec![], Some(uint_token(42))),
```

Comment:
This seems like a tricky case to me. We currently assume numeric literals to always be of type `U256`. In the case of something like...

```
contract Foo:
    pub def bar() -> u128:
        return 42
```

...it seems the compiler should be able to infer that `42` should be of type `u128`. But I'm not sure yet how I would do that. It seems impossible to assign a type directly in `expr_num` without knowing how it is being used. If it is being used as part of a return (e.g. `return 42`) it would be reasonable to lookup the return value of the function but I'm not sure if `expr_num` should try to figure out the context of the usage. So, I'm wondering if `expr_num` should instead just say that this is of numeric type (could be any one at that point) and leave it up to e.g. the `return` to type check it. E.g. we could introduce `Base::AnyNumeric` which would be used as a placeholder for situations like that.

Reply:
Another idea. Instead of introducing a variant such as `AnyNumeric`. We could do something like this:

```
pub enum Base {
    Numeric(Option<NumericSize>),
    Bool,
    Byte,
    Address,
}

#[derive(Clone, Debug, PartialEq, PartialOrd, Ord, Eq)]
pub enum NumericSize {
    U256,
    I256,
    U128,
    I128,
    U64,
    I64
    U32,
    I32
    U16,
    I16,
    U8,
    I8
}
```

So, `expr_num` would resolve to `Base::Numeric(None)` which is to say it is a numeric type but it could be any of them at this point. In practice though, it wouldn't be *any* of them because e.g. if I write `return 40000` then it can't fit into `u8` or `u16` but it could still be anything from `u32` (or `i32`) upward. So, I'm still not sure how exactly this should be dealt with. It feels like what we would like to express in `expr_num` would be a list of allowed types and then leave it to other code to narrow it down.

Anyway, looking forward to discuss that with you after the holidays.

Reply:
```
contract Foo:
    pub def bar() -> u128:
        return 42
```

I think having `42` cast to a `u128` implicitly makes sense, but for now, it might be easier to just require the use of type constructors. So it could instead look like this:

```
contract Foo:
    pub def bar() -> u128:
        return u128(42)
```

In the example above, `42` would be given the type `u256` (as would be the case with any positive literals) and then the return of the constructor would be given type `u128`. With this approach, there's no mystery about the type of an expression, so we don't need to use an optional, which I think would simplify things using the numeric type.

When we get around to introducing implicit casts, we could do something similar to [`move_location`](https://github.com/ethereum/fe/blob/08fcdee7ecebf56d64ef6a104b2a40662c2b48cd/semantics/src/lib.rs#L66), where we would have an attribute like `cast_type: Optional<Type>` that could be set by the expression's context if it needs to be cast to a certain type (in this case `return 42` would cast the return expression to `u128`). We can worry about this later though.

Reply:
Ah, yes, requiring to go through a type constructor strikes me as a pragmatic solution to keep things simple for now :+1:

Reply:
I've put this into its own issue to keep this PR short and focused https://github.com/ethereum/fe/issues/136